ouster-lidar/ouster-ros
# ROS-119: ouster-ros driver automatic reconnection [HUMBLE/IRON/JAZZY]

## Related Issues & PRs
- related #55
- resolves #119
- resolves #233
- related #332
- resolves #337

## Summary of Changes
- Implement sensor reconnection behavior
- Handle sensor reconfiguration from a cold start
- Add a flag to enable sensor reconnect (default disabled)
- properly handle config change outside of the ros driver
- only persist config startup config if persist_config was set
- only `auto_udp` if startup config didn't set an address
- added a parameter to set the dromant period between re-connection attempts
- added a parameter to set the maximum number of subsequent failed re-connection attempts
- added ROS Jazzy to the build :tada:
- added `auto_start` flag

### Known Issues
- Doesn't handle invalid configurations gracefully

## Validation
* Use the driver with default configuration to connect to a sensor & verify normal activity
* Enable the `attempt_reconnect` parameter and verify that it works as expected for the following scenarios
  - Starting from an under powered or unconnected sensor
    - Connect or power up the sensor
    - Verify the driver connects successfully within 10-15 seconds
  - Using a powered up and connected sensor
    - Disconnect the sensor and reconnect it
    - Verify the driver connects successfully within 10-15 seconds
  - Using a powered up and connected sensor
    - power down the sensor and power it up
    - Verify the driver connects successfully within 10-15 seconds (and adjusts to sensor persisted config if different)
  - Using a powered up and connected sensor
    - Use set_config to change sensor configuration
    - Verify the driver reconnects successfully within 10-15 seconds
  - Using a powered up and connected sensor
    - Change some setting on through the sensor configuration page
    - Verify the driver reconnects successfully within 10-15 seconds

### Additional parameters to test
```yaml
# dormant_period_between_reconnects[optional]: wait time in seconds between
# reconnection attempts
dormant_period_between_reconnects: 1.0
# max_failed_reconnect_attempts[optional]: maximum number of attempts trying
# to communicate with the sensor. Counter resets upon successful connection.
max_failed_reconnect_attempts: 2147483647
```
